File: typescript/collatz-conjecture/collatz-conjecture.ts
export function steps(from: number): number {
  if (!Number.isInteger(from) || from <= 0) {
    throw "Only positive integers are allowed";
  }

  return step(from);
}

function step(current_num: number, current_step = 0): number {
  let next_step = current_step + 1;

  if (current_num == 1) {
    return current_step;
  }

  let next_num = current_num % 2 == 0 ? current_num / 2 : 3 * current_num + 1;

  return step(next_num, next_step);
}

File: lua/collatz-conjecture/collatz-conjecture.lua
return function(n)
    assert(n > 0, 'Only positive numbers are allowed')

    local steps = 0

    while n > 1 do
        steps = steps + 1
        if math.fmod(n, 2) == 0 then n = n / 2 else n = 3 * n + 1 end
    end

    return steps
end
